source/io/module_javaScriptInvoke.js

```diff
@@ -358,30 +387,43 @@
                 Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                Module.eggShell.setOccurrenceAsync(occurrencePointer);
                 return;
             }
 
-            var context = undefined;
+            var returnTypeName = getParameterTypeString(returnPointer, 0);
+            var returnValuePointer = undefined;
+            if (returnTypeName !== 'StaticTypeAndData') {
+                // User doesn't want return value. We're passing '*' for the return in VIA code, we get StaticTypeAndData
+                returnValuePointer = JavaScriptInvoke_GetParameterPointer(returnPointer, 0);
+            }
+            if (returnTypeName === 'Array') {
+                returnTypeName += getArrayElementTypeString(returnValuePointer);
+            }
+
+            var currentInvocationState = {
+                completionCallbackRetrieved: false,
+                completionCallbackResolved: 'unresolved'
+            };
+
             var returnValue = undefined;
+            var context = generateContext(occurrencePointer, functionName, returnTypeName, returnValuePointer, errorStatusPointer, errorCodePointer, errorSourcePointer, currentInvocationState);
             try {
                 returnValue = functionToCall.apply(context, parameters);
             } catch (ex) {
                 newErrorStatus = true;
                 newErrorCode = ERRORS.kNIUnableToInvokeAJavaScriptFunction.CODE;
                 newErrorSource = Module.coreHelpers.formatMessageWithException(ERRORS.kNIUnableToInvokeAJavaScriptFunction.MESSAGE + '\'' + functionName + '\'.', ex);
                 Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+                currentInvocationState.completionCallbackResolved = 'rejected';
+                Module.eggShell.setOccurrenceAsync(occurrencePointer);
                 return;
             }
 
-            try {
-                updateReturnValue(functionName, returnPointer, returnValue, errorStatusPointer, errorCodePointer, errorSourcePointer);
-            } catch (ex) {
-                newErrorStatus = true;
-                newErrorCode = ERRORS.kNIUnableToSetReturnValueInJavaScriptInvoke.CODE;
-                newErrorSource = Module.coreHelpers.formatMessageWithException(ERRORS.kNIUnableToSetReturnValueInJavaScriptInvoke.MESSAGE + '\'' + functionName + '\'.', ex);
-                Module.coreHelpers.mergeErrors(newErrorStatus, newErrorCode, newErrorSource, errorStatusPointer, errorCodePointer, errorSourcePointer);
+            if (!currentInvocationState.completionCallbackRetrieved) {
+                updateReturnValue(functionName, returnTypeName, returnValuePointer, returnValue, errorStatusPointer, errorCodePointer, errorSourcePointer);
```

Comment:
Hmm the user could hold onto a reference to `this` and then invoke getCompletionCallback asynchronously, ie:

```javascript
(function () {
var savedContext;

var doEvil = function () {
   var completionCallback = savedContext.getCompletionCallback();
};

window.evilSaveContext = function () {
savedContext = this;
setTimeout(doEvil,0);
};
}());

```

I think the guard for this is to defensively set
currentInvoactionState.completionCallbackRetrieved = true

and maybe the completionCallbackState as well

and a test :)